# Online Retail Data Analysis

## Notebook 1: Initial Data Exploration

### Objective

The goal of this notebook is to understand the raw dataset before making any modifications.

During this notebook we will:

- Load the dataset
- Inspect its structure
- Identify missing values
- Check data types
- Look for duplicates
- Record observations that will guide the data cleaning process

In [2]:
import pandas as pd

In [3]:
df = pd.read_excel("../data/raw/Online Retail.xlsx")

In [4]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


### Observations
This dataset contains 8 columns. These are;
1. InvoiceNo
2. StockCode
3. Description
4. Quantity
5. InvoiceDate
6. UnitPrice
7. CustomerID
8. Country

In [5]:
df.shape

(541909, 8)

### Observations
This dataset contains 541,909 rows and 8 columns 
This dataset is rather large making python a suitable tool for analysis
Each row represents a transaction record

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[us]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 33.1+ MB


### Observations 
Description as 1,454 missing values
CustomerId has 135,080 missing values 
Country is stored as a string
The dataset uses 33.1MBs of storage
InvoiceDate is recognised as a datetime column
Quantity as an integer
Unit price as a decimal number


In [7]:
import pandas as pd
df = pd.read_excel("../data/raw/Online Retail.xlsx")
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

## observations
CustomerID has the highest number of missing value (135,080)
Description has 1454 missing values
The rest of the columns are complete 
Although CustomerID contains many missing values, these rows should not be removed immediately. Missing customer IDs may represent legitimate transactions where customer information was not recorded. Further investigation is required before deciding how to handle these missing values.

In [8]:
import pandas as pd
df = pd.read_excel("../data/raw/Online Retail.xlsx")
df.duplicated().sum()

np.int64(5268)

## Duplicate Records Assessment

### Findings

- The dataset contains **5,268 duplicate rows**.
-

### Initial Assessment

The next step is to inspect the duplicates so that we can know whether they are indeed duplicates or legitimate repeated transactions.

In [9]:

df [df.duplicated()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom
...,...,...,...,...,...,...,...,...
541675,581538,22068,BLACK PIRATE TREASURE CHEST,1,2011-12-09 11:34:00,0.39,14446.0,United Kingdom
541689,581538,23318,BOX OF 6 MINI VINTAGE CRACKERS,1,2011-12-09 11:34:00,2.49,14446.0,United Kingdom
541692,581538,22992,REVOLVER WOODEN RULER,1,2011-12-09 11:34:00,1.95,14446.0,United Kingdom
541699,581538,22694,WICKER STAR,1,2011-12-09 11:34:00,2.10,14446.0,United Kingdom


In [10]:
df [df.duplicated()].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom
587,536412,22273,FELTCRAFT DOLL MOLLY,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom
589,536412,22749,FELTCRAFT PRINCESS CHARLOTTE DOLL,1,2010-12-01 11:49:00,3.75,17920.0,United Kingdom
594,536412,22141,CHRISTMAS CRAFT TREE TOP ANGEL,1,2010-12-01 11:49:00,2.10,17920.0,United Kingdom
598,536412,21448,12 DAISY PEGS IN WOOD BOX,1,2010-12-01 11:49:00,1.65,17920.0,United Kingdom
600,536412,22569,FELTCRAFT CUSHION BUTTERFLY,2,2010-12-01 11:49:00,3.75,17920.0,United Kingdom


In [11]:
df [df.duplicated(keep=False)].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
485,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
489,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
494,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
521,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
548,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom


## Duplicate Records Investigation

### Findings

- The dataset contains **5,268 duplicate rows**.
- Inspection of sample duplicate records showed that all eight columns were identical.
- This suggests the same transaction may have been recorded more than once.

### Decision

Since duplicate records can inflate transaction counts and sales calculations, they should be removed before further analysis.

In [12]:
df_clean = df.drop_duplicates()

In [13]:
df_clean = df.drop_duplicates()

In [14]:
df_clean.shape

(536641, 8)

## Removing Duplicate Records

### Objective

Created a cleaned version of the dataset by removing duplicate records while preserving the original raw dataset.

### Method

A new DataFrame (`df_clean`) was created using `drop_duplicates()`. ensuring that the original dataset remains unchanged and can always be referenced if needed.

In [15]:
df.shape

(541909, 8)

In [16]:
df_clean.shape

(536641, 8)

## Verification

The duplicate removal process was verified by comparing the shapes of the original and cleaned datasets.

- Original dataset: **541,909 rows and 8 columns**
- Cleaned dataset: **536,641 rows and 8 columns**

This confirms that **5,268 duplicate rows** were successfully removed while preserving the original dataset.

In [17]:
df_clean.describe()

,Quantity,InvoiceDate,UnitPrice,CustomerID
count,536641.000000,536641,536641.000000,401604.000000
mean,9.620029,2011-07-04 08:57:06.087421,4.632656,15281.160818
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 10:52:00,1.250000,13939.000000
50%,3.000000,2011-07-19 14:04:00,2.080000,15145.000000
75%,10.000000,2011-10-18 17:05:00,4.130000,16784.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,219.130156,NaN,97.233118,1714.006089


## Descriptive Statistics

### Findings

- The cleaned dataset contains **536,641** transaction records.
- `Quantity`, `InvoiceDate`, and `UnitPrice` contain no missing values.
- `CustomerID` still contains missing values.
- The average quantity purchased per transaction is approximately **9.62** items.
- The average unit price is approximately **4.63**.
- Extremely large positive and negative values exist in both `Quantity` and `UnitPrice`, suggesting that these records require further investigation before deciding whether they are valid transactions or anomalies.

In [18]:
df_clean[df_clean["Quantity"] < 0].head (10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
238,C536391,21980,PACK OF 12 RED RETROSPOT TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
239,C536391,21484,CHICK GREY HOT WATER BOTTLE,-12,2010-12-01 10:24:00,3.45,17548.0,United Kingdom
240,C536391,22557,PLASTERS IN TIN VINTAGE PAISLEY,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
241,C536391,22553,PLASTERS IN TIN SKULLS,-24,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
939,C536506,22960,JAM MAKING SET WITH JARS,-6,2010-12-01 12:38:00,4.25,17897.0,United Kingdom


In [19]:
df_clean[df_clean["Quantity"] < 0][["InvoiceNo", "Quantity"]].head(20)

,InvoiceNo,Quantity
141,C536379,-1
154,C536383,-1
235,C536391,-12
236,C536391,-24
237,C536391,-24
238,C536391,-24
239,C536391,-12
240,C536391,-12
241,C536391,-24
939,C536506,-6


## Investigation of Negative Quantities

### Findings

- Transactions with negative quantities have Invoice Numbers beginning with **`C`**.
- The Unit Prices remain positive.
- Many records share the same CustomerID and InvoiceDate.

### Interpretation

These records appear to represent **credit notes (returns or cancelled transactions)** rather than data entry errors.

### Decision

These records will **not** be removed automatically. Whether they are included or excluded will depend on the business question being answered.

In [20]:
df_clean[df_clean["UnitPrice"] < 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


## Investigation of Negative Unit Prices

### Findings

- Only **2 records** have negative Unit Prices.
- Both records have the description **"Adjust bad debt."**
- Both have a Quantity of **1**.
- Both records have missing Customer IDs.

### Interpretation

These records appear to represent accounting adjustments rather than product sales or customer returns.

### Decision

These records should not be removed automatically. Their inclusion or exclusion depends on the objective of the analysis.

In [21]:
df_clean[df_clean["CustomerID"].isnull ()].shape

(135037, 8)

In [22]:
df_clean[df_clean["CustomerID"].isnull ()].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.00,NaN,United Kingdom
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,2010-12-01 14:32:00,0.85,NaN,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,2010-12-01 14:32:00,1.66,NaN,United Kingdom
1447,536544,21790,VINTAGE SNAP CARDS,9,2010-12-01 14:32:00,1.66,NaN,United Kingdom
1448,536544,21791,VINTAGE HEADS AND TAILS CARD GAME,2,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1449,536544,21801,CHRISTMAS TREE DECORATION WITH BELL,10,2010-12-01 14:32:00,0.43,NaN,United Kingdom
1450,536544,21802,CHRISTMAS TREE HEART DECORATION,9,2010-12-01 14:32:00,0.43,NaN,United Kingdom
1451,536544,21803,CHRISTMAS TREE STAR DECORATION,11,2010-12-01 14:32:00,0.43,NaN,United Kingdom


## Investigation of Missing CustomerID Values

### Findings

- The sampled records have normal Invoice Numbers (they do not begin with `C`).
- Most records represent normal product sales.
- Most sampled transactions are from the United Kingdom.
- Several records belong to the same invoice, indicating multiple products purchased in a single transaction.

### Interpretation

The missing `CustomerID` values do not necessarily indicate poor-quality sales data. They may represent transactions where customer information was not recorded.

### Decision

The records will be retained for overall sales and product analysis. If customer-level analysis is required, only transactions with valid `CustomerID` values will be used.

In [23]:
df_clean["Country"].value_counts ()

Country
United Kingdom          490300
Germany                   9480
France                    8541
EIRE                      8184
Spain                     2528
Netherlands               2371
Belgium                   2069
Switzerland               1994
Portugal                  1510
Australia                 1258
Norway                    1086
Italy                      803
Channel Islands            757
Finland                    695
Cyprus                     611
Sweden                     461
Unspecified                442
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     294
USA                        291
Hong Kong                  284
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


## Transactions by Country

Using `value_counts()`, we found that the United Kingdom has the highest number of transaction records, with 490,300 records. Germany follows with 9,480, while France has 8,541.

The results show that the dataset is heavily concentrated in the United Kingdom.

Note: `value_counts()` counts rows rather than unique invoices, so these figures represent transaction records rather than unique transactions.

In [24]:
df_clean.groupby("Country")["InvoiceNo"].nunique().sort_values(ascending=False)

Country
United Kingdom          23494
Germany                   603
France                    461
EIRE                      360
Belgium                   119
Spain                     105
Netherlands               101
Switzerland                74
Portugal                   71
Australia                  69
Italy                      55
Finland                    48
Sweden                     46
Norway                     40
Channel Islands            33
Japan                      28
Poland                     24
Denmark                    21
Cyprus                     20
Austria                    19
Hong Kong                  15
Unspecified                13
Malta                      10
Singapore                  10
Israel                      9
Iceland                     7
USA                         7
Greece                      6
Canada                      6
Czech Republic              5
European Community          5
Bahrain                     4
Lithuania                   4
Un

## Unique Transactions by Country

We used the number of unique `InvoiceNo` values to compare the number of unique transactions across countries.

The United Kingdom has the highest number of unique invoices, with **23,494**, followed by Germany with **603** and France with **461**.

This confirms that the United Kingdom dominates the dataset in terms of both transaction records and unique invoices.

The large difference between UK transaction records (490,300) and unique invoices (23,494) indicates that individual invoices contain multiple product records.